### PR TITLE
♻️ Refactor Waitingroom

### PR DESCRIPTION
**Pull request by SteffenKn**
**Changes:**

1. Remove Waiting Room
2. Add Live Execution Tracker
3. Fix handling parallel UserTasks and ManualTasks
4. Remove EventListeners in deactivate
5. Adjust existing E2E tests
6. Fix version of bpmn-js-custom-bundle
7. Adjust interfaces
8. Upgrade Management API to 1.5.0
9. Upgrade ProcessEngine Runtime to 4.0.0

**Issues:**

PR: #1152 

## How can others test the changes?

- Open up the BPMN-Studio.
- Go to the Detail View of any diagram.
- Deploy and run the diagram.
- **Notice that the lice execution tracker will now be shown.**
- Click on a Manual or UserTask if existing.

## PR-Checklist

Please check the boxes in this list after submitting your PR:

- [x] You can merge this PR **right now** (if not, please prefix the title with "WIP: ")
- [x] I've tested **all** changes included in this PR.
- [x] I've also reviewed this PR myself before submitting (e.g. for scrambled letters, typos, etc.).
- [x] I've rebased the `develop` branch with my branch before finishing this PR.
- [x] I've **summarized all changes** in a list above.
- [x] I've mentioned all **PRs, which relate to this one**.
- [x] I've prefixed my Pull Request title is according to [gitmoji guide](https://gitmoji.carloscuesta.me/).

## Example

<details>
<summary>
:warning: Before merging please click here to expand and follow the guide.
</summary>
<br>

Please use `:twisted_rightwards_arrows:` at the beginning of your merge commit title.

Example 1:

<pre>
<code>
:twisted_rightwards_arrows: :bug: Fix Wrong Text Decoration at ...
</code>
</pre>

To get your commit message, just copy the first part of this pull request.

Example 2:
<pre>
<code>
**Changes:**

- Fixes Wrong Text Decoration at ...
- Fixes some typos
- ...

**Issues:**

Closes #NumberOfFixedIssue

PR: #NumberOfThisPR
</code>
</pre>
</details>

## Emojis

<details>
<summary>
Expand for a list of most used Emojis.
</summary>
<br>

Please prefix your commit messages with an Emoji.

Ref: https://gitmoji.carloscuesta.me/

| Description              | Glyphe               | Emoji  |
|--------------------------|----------------------|--------|
| Bugfix                   | `:bug:`              | 🐛     |
| Configuration releated   | `:wrench:`           | 🔧     |
| Cosmetic                 | `:lipstick:`         | 💄     |
| Dependencies Downgrade   | `:arrow_down:`       | ⬇️     |
| Dependencies Upgrade     | `:arrow_up:`         | ⬆️     |
| Formatting               | `:art:`              | 🎨     |
| Improving Performance    | `:zap:`              | ⚡️     |
| Initial commit           | `:tada:`             | 🎉     |
| Linter                   | `:rotating_light:`   | 🚨     |
| Miscellaneous            | `:package:`          | 📦     |
| New Feature              | `:sparkles:`         | ✨     |
| Refactoring Code         | `:recycle:`          | ♻️     |
| Releasing / Version tags | `:bookmark:`         | 🔖     |
| Removing Stuff           | `:fire:`             | 🔥     |
| Tests                    | `:white_check_mark:` | ✅     |
| Work In Progress (WIP)   | `:construction:`     | 🚧     |

</details>

### CODE REVIEW

File: src/modules/design/diagram-detail/diagram-detail.ts [hunk 1]
```diff
@@ -51,6 +51,10 @@ export class DiagramDetail {
   private _ipcRenderer: any;
   private _solutionService: ISolutionService;
   private _managementApiClient: IManagementApi;
+  private _ipcRendererEventListeners: Array<{
```
Comment (ElRaptorus) [inline src/modules/design/diagram-detail/diagram-detail.ts]: I think moving 
```ts
{
  name: string,
  function: Function,
}
```
to a dedicated type would improve readability here.

File: src/app.ts [hunk 2]
```diff
@@ -181,7 +181,7 @@ export class App {
         nav: true,
       },
       {
-        route: ['processdef/:processModelId/task/:taskId/dynamic-ui'],
+        route: ['correlation/:correlationId/processdef/:processModelId/task/:taskId/dynamic-ui'],
```
[the remark this answers is not in the document]
Comment (SteffenKn) [reply src/app.ts]:
```suggestion
        route: ['correlation/:correlationId/processmodel/:processModelId/task/:taskId/dynamic-ui'],
```

File: src/modules/live-execution-tracker/live-execution-tracker.ts [hunk 3]
```diff
@@ -0,0 +1,490 @@
...
+      return;
+    }
+
+    this._modeling.setColor(elementsWithColor, {
+      stroke: defaultBpmnColors.none.border,
+      fill: defaultBpmnColors.none.fill,
+    });
+  }
+
+  private async _importXml(modeler: IBpmnModeler, xml: string): Promise<void> {
```
[the remark this answers is not in the document]
Comment (SteffenKn) [reply src/modules/live-execution-tracker/live-execution-tracker.ts]: @cmg-dev

File: package.json [hunk 4]
```diff
@@ -92,45 +92,45 @@
...
-    "@process-engine/management_api_client": "1.2.0",
-    "@process-engine/management_api_contracts": "1.2.1",
-    "@process-engine/process_engine_runtime": "3.2.1",
+    "@essential-projects/errors_ts": "1.4.2",
+    "@essential-projects/http_contracts": "2.3.3",
+    "@process-engine/consumer_api_contracts": "1.3.0",
+    "@process-engine/kpi_api_contracts": "1.1.0",
+    "@process-engine/management_api_client": "1.4.0",
+    "@process-engine/management_api_contracts": "1.4.0",
+    "@process-engine/process_engine_runtime": "4.0.0-a581766b-b146",
```
Comment (Paulomart) [inline package.json]: Can this reference updated to a release?

File: src/contracts/bpmnmodeler/IOverlayManager.ts [hunk 5]
```diff
@@ -0,0 +1,11 @@
+import {IOverlayDescriptor} from './IOverlayDescriptor';
+import {IOverlays} from './IOverlays';
+import {IShape} from './IShape';
+
+export interface IOverlayManager {
+  _overlays: IOverlays;
```
Comment (Paulomart) [inline src/contracts/bpmnmodeler/IOverlayManager.ts]: Why expose a field like that? Better add methods to manipulate the field.

File: src/contracts/bpmnmodeler/IOverlays.ts [hunk 6]
```diff
@@ -0,0 +1,5 @@
+import {IOverlay} from './IOverlay';
+
+export interface IOverlays {
```
Comment (Paulomart) [inline src/contracts/bpmnmodeler/IOverlays.ts]: Looks like a [Map](https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/Map) to me.

File: src/contracts/bpmnmodeler/IViewbox.ts [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+  width: number;
+  height: number;
+  x: number;
+  y: number;
+  scale: number;
+  inner: IInnerViewbox;
+  outer: IOuterViewbox;
+}
+
+export interface IInnerViewbox {
```
Comment (Paulomart) [inline src/contracts/bpmnmodeler/IViewbox.ts]: Please limit the amount of interfaces to 1 per file.

File: src/contracts/bpmnmodeler/IViewbox.ts [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+  x: number;
+  y: number;
+  width: number;
+  height: number;
+}
+
+export interface IOuterViewbox {
```
Comment (Paulomart) [inline src/contracts/bpmnmodeler/IViewbox.ts]: Please limit the amount of interfaces to 1 per file.

File: src/modules/design/diagram-detail/diagram-detail.ts [hunk 9]
```diff
@@ -23,6 +23,11 @@ interface RouteParameters {
   diagramName?: string;
 }
 
+type EventListener = {
```
Comment (Paulomart) [inline src/modules/design/diagram-detail/diagram-detail.ts]: I would suggest making this an interface and using the prefix `IPC`

File: src/contracts/bpmnmodeler/IOverlayManager.ts [hunk 10]
```diff
@@ -0,0 +1,11 @@
+import {IOverlayDescriptor} from './IOverlayDescriptor';
+import {IOverlays} from './IOverlays';
+import {IShape} from './IShape';
+
+export interface IOverlayManager {
```
Comment (Paulomart) [inline src/contracts/bpmnmodeler/IOverlayManager.ts]: `IOverlayService` ?

File: src/modules/live-execution-tracker/live-execution-tracker.ts [hunk 11]
```diff
@@ -0,0 +1,520 @@
+
```
Comment (Paulomart) [inline src/modules/live-execution-tracker/live-execution-tracker.ts]: Remove empty line.

File: src/modules/live-execution-tracker/live-execution-tracker.ts [hunk 12]
```diff
@@ -0,0 +1,520 @@
...
+import {inject} from 'aurelia-framework';
+import {Router} from 'aurelia-router';
+
+import {IIdentity} from '@essential-projects/iam_contracts';
+import * as bundle from '@process-engine/bpmn-js-custom-bundle';
+
+import {Correlation, CorrelationProcessModel, IManagementApi, TokenHistoryEntry} from '@process-engine/management_api_contracts';
```
Comment (alexanderkasten) [inline src/modules/live-execution-tracker/live-execution-tracker.ts]:
```suggestion
import {
  Correlation, 
  CorrelationProcessModel,
  IManagementApi, 
  TokenHistoryEntry,
} from '@process-engine/management_api_contracts';
```

File: src/contracts/bpmnmodeler/IOverlays.ts (hunk 6 above)
Comment (cmg-dev) [reply src/contracts/bpmnmodeler/IOverlays.ts]: Yeah, this would be better.

File: package.json (hunk 4 above)
Comment (SteffenKn) [reply package.json]: Not yet

File: src/modules/live-execution-tracker/live-execution-tracker.ts [hunk 13]
```diff
@@ -0,0 +1,520 @@
...
+    await this._importXml(this._diagramModeler, xml);
+
+    /*
+     * Remove all colors if the diagram has already colored elements.
+     * For example, if the user has some elements colored orange and is running the diagram, one would think in LiveExecutionTracker that the element
+     *  is active although it is not active.
+    */
+    this._clearColors();
+
+    // Get all Elements that can have a token
```
Comment (alexanderkasten) [inline src/modules/live-execution-tracker/live-execution-tracker.ts]:
```suggestion
    // Get all elements that can have a token
```

File: src/modules/design/diagram-detail/diagram-detail.ts (hunk 9 above)
Comment (cmg-dev) [reply src/modules/design/diagram-detail/diagram-detail.ts]: I oppose this.

This has virtually no use in the contracts.

This is okay, as it is.

File: src/modules/live-execution-tracker/live-execution-tracker.ts [hunk 14]
```diff
@@ -0,0 +1,520 @@
...
+  IModdleElement,
+  IModeling,
+  IOverlayManager,
+  IShape,
+  NotificationType,
+} from '../../contracts/index';
+import environment from '../../environment';
+import {NotificationService} from '../notification/notification.service';
+
+interface RouteParameters {
```
Comment (cmg-dev) [inline src/modules/live-execution-tracker/live-execution-tracker.ts]: why not a type for this?

File: src/contracts/bpmnmodeler/IOverlayManager.ts (hunk 5 above)
Comment (SteffenKn) [reply src/contracts/bpmnmodeler/IOverlayManager.ts]: The "OverlayManage" is part of BPMN-JS.

File: src/modules/live-execution-tracker/live-execution-tracker.ts (hunk 11 above)
Comment (SteffenKn) [reply src/modules/live-execution-tracker/live-execution-tracker.ts]:
```suggestion
```

File: src/modules/live-execution-tracker/live-execution-tracker.ts [hunk 15]
```diff
@@ -0,0 +1,490 @@
...
+    this._startPolling();
+  }
+
+  public detached(): void {
+    this._attached = false;
+    this._stopPolling();
+  }
+
+  private async _colorizeXml(xml: string): Promise<string> {
+    await this._importXml(this._diagramModeler, xml);
```
[the remark this answers is not in the document]
Comment (cmg-dev) [reply src/modules/live-execution-tracker/live-execution-tracker.ts]: Yes,

implement two functions:

1. `importXmlFromDiagramViewer()`
1. `importXmlFromDiagramModeler()`

File: src/modules/live-execution-tracker/live-execution-tracker.ts [hunk 16]
```diff
@@ -0,0 +1,520 @@
...
+  }
+
+  private async _colorizeXml(xml: string): Promise<string> {
+    // Import the xml to the modeler to add colors to it
+    await this._importXml(this._diagramModeler, xml);
+
+    /*
+     * Remove all colors if the diagram has already colored elements.
+     * For example, if the user has some elements colored orange and is running the diagram, one would think in LiveExecutionTracker that the element
+     *  is active although it is not active.
```
Comment (cmg-dev) [inline src/modules/live-execution-tracker/live-execution-tracker.ts]: please 80 Chars per line...

File: src/modules/live-execution-tracker/live-execution-tracker.ts [hunk 17]
```diff
@@ -0,0 +1,519 @@
...
+
+      document.getElementById(element.id).addEventListener('click', this._handleElementClick);
+
+      this._elementsWithEventListeners.push(element.id);
+    }
+  }
+
+  private _handleElementClick: (event: MouseEvent) => void =
+    (event: MouseEvent): void => {
+      const elementId: string = (event.target as any).id;
```
Comment (alexanderkasten) [inline src/modules/live-execution-tracker/live-execution-tracker.ts]: you know the type of the target element..

```suggestion
      const elementId: string = (event.target as HTMLDivElement).id;
```

File: src/modules/live-execution-tracker/live-execution-tracker.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (SteffenKn) [reply src/modules/live-execution-tracker/live-execution-tracker.ts]: @cmg-dev Ok. Same for export i guess.

File: src/modules/design/diagram-detail/diagram-detail.ts (hunk 9 above)
Comment (Paulomart) [reply src/modules/design/diagram-detail/diagram-detail.ts]: EventListener is a global object in JS. Overriding is not to nice...

**Pull request merged**